apps/labs/posts/automated-accessibility-testing-with-bokeh.md

```diff
@@ -0,0 +1,333 @@
+keeping things balanced, and rendering.
+
+Here's a picture from the Bokeh docs on [contributing to
+BokehJs](https://docs.bokeh.org/en/latest/docs/dev_guide/bokehjs.html#contributor-guide-bokehjs)
+to illustrate how data flows:
+
+<img alt="Image of Bokeh data model flow represented as a flow chart moving from
+Python to JSON, to Javascript, and then to final HTML output"
+src="/posts/automated-accessibility-testing-with-bokeh/bokeh_bokehjs.jpg"
+style={{padding: '10px 20px'}}/>
+
+Now that I knew the inner workings and where things fit, I had an idea of how to
+start the testing. This, however, brought up the question of "what" to test
+things with.
+
+## Accessibility testing tools
+
+For the right blend of power, ease of use, and modernity, my mentor
+[Gabriel](https://github.com/gabalafou) and I decided to use
+[Playwright](https://playwright.dev/) for setting up the testing environment.
+
+I had hoped that using Playwright would help me wriggle out of a particular
+problem. You see, sometimes when I'm uncertain of a direction to go, I let
+someone/something else be the deciding factor. Remember I mentioned that Bokeh
+had both Javascript and Python equivalents for the same thing? Well, I hoped
+Playwright would be opinionated about whether to test with Javascript or Python.
+In an interesting series of events, I realized that it had support for both.
+
+We also decided that having [axe-core](https://github.com/dequelabs/axe-core) in
+the tests might prove useful. Axe-core is a tool that takes in a HTML node and
+scans it against the [Axe
+API](https://www.deque.com/axe/core-documentation/api-documentation/) for known
+Accessibility issues.
+
+I ended up starting with Python because it was closer to the intended usage of
+Bokeh (as a Python tool) and had little friction for me, but after about a week
+of development, there were doubts on if it would be the better approach going
+forward. This was because generated plots and visualizations would be rendered
+in the browser and the code responsible for that would be in Javascript.
+Additionally, although the testing tools had Python equivalents, Python wasn't a
+first-class candidate for Playwright.
+
+Javascript had first-class support for the testing tools and it was the language
+responsible for interactive rendering on the browser, so we ported to that. It
+was just before switching to Javascript equivalents that I made an important
+discovery: Due to the very nature of Bokeh, axe-core would be useless for lots
+of plots. Remember I mentioned that despite generating HTML, Bokeh plots could
+not be treated in the same way as web pages? Well, web pages tend to have a
+structure, and elements with semantic meaning. However, most Bokeh data
+visualizations end up getting wrapped in an HTML `<canvas>` with some styling in
+the [Shadow
+DOM](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM),
+which is something axe does not know how to work with. The Bokeh data
```

Comment:
This could be confusing to anyone who knows Axe-core well, because it actually does support Shadow DOM. Canvas is the issue, not shadow DOM.

```suggestion
visualizations end up getting wrapped in an HTML `<canvas>`, which is something axe does not know how to work with. The Bokeh data
```